From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Max Filippov <jcmvbkbc@gmail.com>
Date: Fri, 31 Jan 2020 16:38:02 -0800
Subject: [PATCH 09/73] drivers/mtd: add missing _xiptext/__always_inline
 annotations

Add _xiptext/__always_inline annotations to functions used in ??? to fix
XIP kernel crashes.

Signed-off-by: Max Filippov <jcmvbkbc@gmail.com>
(cherry picked from commit 972f4cc0eb8666d6281782faf72e566475c8c30f)
---
 drivers/mtd/chips/cfi_util.c | 12 ++++++------
 include/linux/mtd/cfi.h      |  4 ++--
 include/linux/mtd/map.h      |  2 +-
 3 files changed, 9 insertions(+), 9 deletions(-)

diff --git a/drivers/mtd/chips/cfi_util.c b/drivers/mtd/chips/cfi_util.c
index ef0aa6890bc06..6d04331ac0421 100644
--- a/drivers/mtd/chips/cfi_util.c
+++ b/drivers/mtd/chips/cfi_util.c
@@ -36,8 +36,8 @@ EXPORT_SYMBOL(cfi_udelay);
 /*
  * Returns the command address according to the given geometry.
  */
-uint32_t cfi_build_cmd_addr(uint32_t cmd_ofs,
-				struct map_info *map, struct cfi_private *cfi)
+uint32_t __xipram cfi_build_cmd_addr(uint32_t cmd_ofs,
+				     struct map_info *map, struct cfi_private *cfi)
 {
 	unsigned bankwidth = map_bankwidth(map);
 	unsigned interleave = cfi_interleave(cfi);
@@ -63,7 +63,7 @@ EXPORT_SYMBOL(cfi_build_cmd_addr);
  * It looks too long to be inline, but in the common case it should almost all
  * get optimised away.
  */
-map_word cfi_build_cmd(u_long cmd, struct map_info *map, struct cfi_private *cfi)
+map_word __xipram cfi_build_cmd(u_long cmd, struct map_info *map, struct cfi_private *cfi)
 {
 	map_word val = { {0} };
 	int wordwidth, words_per_bus, chip_mode, chips_per_word;
@@ -199,9 +199,9 @@ EXPORT_SYMBOL(cfi_merge_status);
  * If prev_val is non-null, it will be set to the value at the command address,
  * before the command was written.
  */
-uint32_t cfi_send_gen_cmd(u_char cmd, uint32_t cmd_addr, uint32_t base,
-				struct map_info *map, struct cfi_private *cfi,
-				int type, map_word *prev_val)
+uint32_t __xipram cfi_send_gen_cmd(u_char cmd, uint32_t cmd_addr, uint32_t base,
+				   struct map_info *map, struct cfi_private *cfi,
+				   int type, map_word *prev_val)
 {
 	map_word val;
 	uint32_t addr = base + cfi_build_cmd_addr(cmd_addr, map, cfi);
diff --git a/include/linux/mtd/cfi.h b/include/linux/mtd/cfi.h
index 35ca19ae21ae0..86d473f00363e 100644
--- a/include/linux/mtd/cfi.h
+++ b/include/linux/mtd/cfi.h
@@ -304,7 +304,7 @@ uint32_t cfi_send_gen_cmd(u_char cmd, uint32_t cmd_addr, uint32_t base,
 				struct map_info *map, struct cfi_private *cfi,
 				int type, map_word *prev_val);
 
-static inline uint8_t cfi_read_query(struct map_info *map, uint32_t addr)
+static __always_inline uint8_t cfi_read_query(struct map_info *map, uint32_t addr)
 {
 	map_word val = map_read(map, addr);
 
@@ -320,7 +320,7 @@ static inline uint8_t cfi_read_query(struct map_info *map, uint32_t addr)
 	return cfi32_to_cpu(map, val.x[0]);
 }
 
-static inline uint16_t cfi_read_query16(struct map_info *map, uint32_t addr)
+static __always_inline uint16_t cfi_read_query16(struct map_info *map, uint32_t addr)
 {
 	map_word val = map_read(map, addr);
 
diff --git a/include/linux/mtd/map.h b/include/linux/mtd/map.h
index 75b0b2abc8806..d2e8edec4e7fe 100644
--- a/include/linux/mtd/map.h
+++ b/include/linux/mtd/map.h
@@ -388,7 +388,7 @@ static inline map_word map_word_ff(struct map_info *map)
 	return r;
 }
 
-static inline map_word inline_map_read(struct map_info *map, unsigned long ofs)
+static __always_inline map_word inline_map_read(struct map_info *map, unsigned long ofs)
 {
 	map_word r;
 
